doc/design/prefix_tree.md

```diff
@@ -0,0 +1,482 @@
+
+### How to ensure Relocatability
+
+1. Use members that are trivially-copyable.
+1. Only refer to memory owned by the object.
+1. Only use indices or offsets relative to the start of the object (`this`).
+
+**Remark:** classes are relocatable members are relocatable (trivially-copyable) if and only if
+all its members are relocatable. All primitive data types are trivially-copyable.
+
+**Claim:** `PrefixTree` is relocatable.
+
+This follows since we use only `relocatable_ptr`s in the nodes and the remaining members are of type `TypedAllocator`,
+which is trivially-copyable. The latter holds since it uses indices internally and owns its memory.
+
+## Memory consumption
+
+Since we need to be prepared for a maximum number of values C, we need to have a sufficient number of nodes to accomodate them.
+This is trivial for the data nodes, which we need a most C of as we only store at most C values in the tree.
+Each of the nodes just has an extra pointer, which is fairly minor.
+
+It is more problematic for the nodes used for the individual characters, which require 3 pointers for each character stored.
+In general there are ways to compress sequences of characters in the tree where no branching occurs,
+but this is more complex and not feasible with static memory (how many chars should we store in a node?).
+
+Tightly-estimating there worst case number of thse nodes is harder, since in general we may have keys with little prefix overlap.
+By the pigeonhole principle, we will always have some overlap if C is large enough.
+For example we only have 256 character values and hence for C > 256 values with different keys
+we will have at least one shared prefix (of one character).
+This argument becomes increasingly weaker for subsequent levels of the tree.
+
+Disregarding this, a worst case estimate is
+
+```text
+#Nodes <= MaxKeyLength * C.
+```
+
+This can be imrpoved somewhat for C > 256 to
+
+```text
+#Nodes <= (MaxKeyLength-1) * C + 256.
+```
+
+which does not help much. Further improved estimates accounting for more upper-level overlap appear not really useful.
+
+This leads to considerable memory consumption for each object, but at runtime this memory is only needed for pathological cases
+which have little prefix overlap in the keys and long key strings.
+
+We can introduce another control parameter which lets the user limit the number of these key nodes,
+but it is hard to quantify its effects for the user.
+
+If we consider the MaxKeyLength a constant, the memory consumption is still linear in C.
+
+## Generalization
+
+The `TypedAllocator` and `relocatable_ptr` can be ued to create arbitrary relocatable linked structures,
+e.g. lists, trees of different kinds like k-d-trees, maps, sets etc.
+
+These can be used in shared memory and transferred with zero-copy overhead (once constructed) using iceoryx.
+
+## Alternatives
+
+### Brute-force
+
+* arrays work well enough up to some point (estimate in 100-300 strings), depending on the searches
```

Comment:
Maybe also in the 1000s, but 10000s are probably too much for brute force search.